File: dataModule/src/main/java/com/mobnetic/coinguardian/model/market/BitcoinToYou.java
package com.mobnetic.coinguardian.model.market;

import java.util.HashMap;
import java.util.LinkedHashMap;

import org.json.JSONObject;

import com.mobnetic.coinguardian.model.CheckerInfo;
import com.mobnetic.coinguardian.model.Market;
import com.mobnetic.coinguardian.model.Ticker;
import com.mobnetic.coinguardian.model.currency.Currency;
import com.mobnetic.coinguardian.model.currency.VirtualCurrency;

public class BitcoinToYou extends Market {

	private final static String NAME = "BitcoinToYou";
	private final static String TTS_NAME = "Bitcoin To You";
	private final static String URL = "https://back.bitcointoyou.com/api/ticker?pair=%1$s_%2$sC";
	private final static HashMap<String, CharSequence[]> CURRENCY_PAIRS = new LinkedHashMap<String, CharSequence[]>();
	static {
		CURRENCY_PAIRS.put(VirtualCurrency.BTC, new String[]{
				Currency.BRL
			});
	}
	
	public BitcoinToYou() {
		super(NAME, TTS_NAME, CURRENCY_PAIRS);
	}

	@Override
	public String getUrl(int requestId, CheckerInfo checkerInfo) {
/*
		String pairId = checkerInfo.getCurrencyPairId();
		if (pairId == null) {
			pairId = String.format("%1$s_%2$s", checkerInfo.getCurrencyBaseLowerCase(), checkerInfo.getCurrencyCounterLowerCase());
		}
 */
		String result = String.format(URL, checkerInfo.getCurrencyBase(), checkerInfo.getCurrencyCounter());
		return result;
	}
	
	@Override
	protected void parseTickerFromJsonObject(int requestId, JSONObject jsonObject, Ticker ticker, CheckerInfo checkerInfo) throws Exception {
		final JSONObject tickerJsonObject = jsonObject.getJSONObject("summary");
//		ticker.bid = tickerJsonObject.getDouble("buy");
//		ticker.ask = tickerJsonObject.getDouble("sell");
		ticker.vol = tickerJsonObject.getDouble("amount");
		ticker.high = tickerJsonObject.getDouble("high");
		ticker.low = tickerJsonObject.getDouble("low");
		ticker.last = tickerJsonObject.getDouble("last");
	}
}

File: dataModule/src/main/java/com/aneonex/bitcoinchecker/datamodule/model/market/BitcoinToYou.kt
package com.aneonex.bitcoinchecker.datamodule.model.market

import com.aneonex.bitcoinchecker.datamodule.exceptions.MarketParseException
import com.aneonex.bitcoinchecker.datamodule.model.CheckerInfo
import com.aneonex.bitcoinchecker.datamodule.model.CurrencyPairInfo
import com.aneonex.bitcoinchecker.datamodule.model.Ticker
import com.aneonex.bitcoinchecker.datamodule.model.market.generic.SimpleMarket
import com.aneonex.bitcoinchecker.datamodule.util.forEachJSONObject
import org.json.JSONArray
import org.json.JSONObject

class BitcoinToYou : SimpleMarket(
    "BitcoinToYou",
    "https://back.bitcointoyou.com/api/v2/pair/getPairList",
    "https://back.bitcointoyou.com/api/v2/ticker?pair=%1\$s",
    "Bitcoin To You"
) {

    override fun getPairId(checkerInfo: CheckerInfo): String? {
        return if (checkerInfo.currencyPairId == null)
                    with(checkerInfo) { "${currencyBase}_${currencyCounter}C" }
                else
                    super.getPairId(checkerInfo)
    }

    override fun parseCurrencyPairs(
        requestId: Int,
        responseString: String,
        pairs: MutableList<CurrencyPairInfo>
    ) {
        fun getDisplaySymbol(symbolString: String): String {
            @Suppress("SpellCheckingInspection")
            return when(symbolString){
                "CBRL" -> "BRL"
                "BRLC" -> "BRL"
                else -> symbolString
            }
        }

        val data = JSONArray(responseString)
        data.forEachJSONObject {
            val symbol = it.getString("symbol")
            val assets = symbol.split('_')
            if(assets.size == 2) {
                pairs.add(
                    CurrencyPairInfo(
                        getDisplaySymbol(assets[0]),
                        getDisplaySymbol(assets[1]),
                        symbol
                    )
                )
            }
        }
    }

    @Throws(Exception::class)
    override fun parseTickerFromJsonObject(requestId: Int, jsonObject: JSONObject, ticker: Ticker, checkerInfo: CheckerInfo) {
        jsonObject.getJSONObject("summary").also { data ->
            ticker.vol = data.getDouble("amount")
            if(ticker.vol <= 0)
                throw MarketParseException("No trading volume")

            ticker.volQuote = data.getDouble("quote_volume")
            ticker.high = data.getDouble("high")
            ticker.low = data.getDouble("low")
            ticker.last = data.getDouble("last")
        }
    }
}